lib/rules/no-constant-condition.js

```diff
@@ -129,7 +129,7 @@ module.exports = {
                     const isLeftConstant = isConstant(node.left, inBooleanPosition);
                     const isRightConstant = isConstant(node.right, inBooleanPosition);
                     const isLeftShortCircuit = (isLeftConstant && isLogicalIdentity(node.left, node.operator));
-                    const isRightShortCircuit = (isRightConstant && isLogicalIdentity(node.right, node.operator));
+                    const isRightShortCircuit = (inBooleanPosition && isRightConstant && isLogicalIdentity(node.right, node.operator));
```

Comment:
#11308 did not handle following code:
```js
if ((foo || true) === 'baz') {}
```

[demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tY29uc3RhbnQtY29uZGl0aW9uOiBcImVycm9yXCIgKi9cblxuaWYgKChmb28gfHwgdHJ1ZSkgPT09ICdiYXonKSB7fSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnsic29ydC1pbXBvcnRzIjoyfSwiZW52Ijp7ImVzMjAyMCI6dHJ1ZX19fQ==)

Reply:
This is indeed a bug that should be fixed whether or not #13238 gets accepted.

Could we maybe also this:

```js
(
    !node.parent ||
    node.parent.type !== "BinaryExpression" ||
    !(EQUALITY_OPERATORS.includes(node.parent.operator) || RELATIONAL_OPERATORS.includes(node.parent.operator))
)
```

replace with just `inBooleanPosition`?

I think it might fix similar bugs, such as:

```js
/*eslint no-constant-condition: "error"*/

if ((foo || 'bar' || 'bar') === 'bar'); // error, false positive
```

Though, it seems that the whole condition starting with  `node.operator === "||" &&` might be removed in this PR (if #13238 gets accepted), but we should also doublecheck what's happening with the recursion in `isLogicalIdentity`. It already looks buggy in the actual version:

```js
/*eslint no-constant-condition: "error"*/

if (a && false || b); // error, false positive
```

Reply:
The cause is an inappropriate  propagation of short circuit.
fixed in https://github.com/eslint/eslint/pull/13245/commits/96d083497afb3bd8fc68b5d87b64579070565228